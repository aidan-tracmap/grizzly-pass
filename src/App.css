.App-header {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  height: 50px;
}

.App-sideMenu {
  position: fixed;
  top: 50px;
  left: 0;
  bottom: 0;
  width: 60px;
}

.App-content {
  display: flex;
  flex-direction: row;
  margin-top: 50px;
  margin-left: 60px;
  padding: 30px;
}

.App-month {
  padding-right: 30px;
}

.App-month:last-child {
  padding-right: 0;
}

.App-monthTitle {
  font-size: 23px;
  font-weight: 300;
  padding-bottom: 10px;
}
