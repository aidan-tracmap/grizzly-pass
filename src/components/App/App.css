.App {
  position: absolute;
  width: 100%;
  min-height: 100%;
}

.App-header {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  height: 50px;
  z-index: 1;
}

.App-sideMenu {
  position: fixed;
  top: 50px;
  left: 0;
  bottom: 0;
  width: 60px;
  z-index: 1;
}

.App-footer {
  background: #f2f3f4;
  margin-left: 80px;
}

.App-content {
  display: flex;
  flex-direction: row;
  margin-top: 50px;
  margin-left: 60px;
  padding: 30px 30px 80px 30px;
}

.App-month {
  padding-right: 30px;
}

.App-month:last-child {
}

.App-monthTitle {
  font-size: 23px;
  font-weight: 300;
  padding-bottom: 10px;
}

.App-cardWrapper {
  padding-bottom: 10px;
}
