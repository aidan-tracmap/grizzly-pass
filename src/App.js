import React from "react";
import Header from "./Header";
import SideMenu from "./SideMenu";
import "./App.css";

export default function App() {
  return (
    <div className="App">
      <div className="App-sideMenu">
        <SideMenu />
      </div>
      <div className="App-header">
        <Header />
      </div>
      <p className="App-content">
        https://grizzly-pass.surge.sh
      </p>
    </div>
  );
}
