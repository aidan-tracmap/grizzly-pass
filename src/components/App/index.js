import React, { PropTypes } from "react";
import moment from "moment";
import Header from "../Header";
import SideMenu from "../SideMenu";
import Card from "../Card";
import Legend from "../Legend";
import { statusIds } from "../../types";
import "./App.css";

export default function App({ onFileDrop, title = "", projects = [] }) {
  const months = splitIntoMonths(projects).map(({ month, projects }) => {
    const cards = projects.map(project => (
      <Card
        key={project.id}
        title={project.title}
        person={project.person}
        time={project.time}
        progress={project.progress}
        status={project.status}
        labels={project.labels}
      />
    ));
    return (
      <div key={month} className="App-month">
        <h2 className="App-monthTitle">{month}</h2>
        {cards}
      </div>
    );
  });

  const hasProjects = projects && projects.length;

  return (
    <div
      className="App"
      onDrop={e => handleDrop(onFileDrop, e)}
      onDragOver={handleDragOver}
    >
      <div className="App-sideMenu">
        <SideMenu />
      </div>
      <div className="App-header">
        <Header title={title} />
      </div>
      <div className="App-content">{months}</div>
      {hasProjects &&
        <div className="App-footer">
          <Legend projects={projects} />
        </div>}
    </div>
  );
}

App.propTypes = {
  onFileDrop: PropTypes.func.isRequired,
  projects: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      person: PropTypes.string.isRequired,
      time: PropTypes.string.isRequired,
      progress: PropTypes.number.isRequired,
      status: PropTypes.oneOf(statusIds).isRequired,
      labels: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.string.isRequired,
          initial: PropTypes.string.isRequired,
          colour: PropTypes.string.isRequired,
          title: PropTypes.string.isRequired
        })
      )
    })
  ),
  title: PropTypes.string
};

function splitIntoMonths(projects) {
  const monthIndexes = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11];
  return monthIndexes
    .map(monthIndex =>
      projects.filter(project => moment(project.time).month() === monthIndex))
    .filter(month => month.length > 0)
    .map(month => ({
      month: moment(month[0].time).format("MMMM"),
      projects: month
    }));
}

function handleDragOver(e) {
  e.preventDefault();
}

function handleDrop(onFileDrop, e) {
  e.preventDefault();
  const dt = e.dataTransfer;
  if (dt.items) {
    const item = Array.prototype.find.call(
      dt.items,
      ({ kind }) => kind === "file"
    );
    if (item) onFileDrop(item.getAsFile());
  } else if (dt.files.length) {
    onFileDrop(dt.files[0]);
  }
}
