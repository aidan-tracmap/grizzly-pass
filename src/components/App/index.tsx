import * as React from "react";
import * as moment from "moment";
import Header from "../Header";
import SideMenu from "../SideMenu";
import Card from "../Card";
import Legend from "../Legend";
import { TabId, Status, statusIds, tabIds } from "../../types";
import "./App.css";

interface Project {
  id: string;
  title: string;
  person: string;
  time: string;
  progress: number;
  status: Status;
  labels?: {
    id: string;
    initial: string;
    colour: string;
    title: string;
  }[];
}

export interface AppProps {
  onFileDrop: (file: File) => void;
  title?: string;
  selectedTab?: TabId;
  onTabChange: (tabId: TabId) => void;
  errorMessage?: string | string[];
  projects?: Project[];
}

export default function App(
  {
    onFileDrop,
    title = "",
    projects = [],
    errorMessage,
    selectedTab,
    onTabChange
  }: AppProps
) {
  const months = splitIntoMonths(projects).map(({ month, projects }) => {
    const cards = projects.map(project => (
      <div className="App-cardWrapper" key={project.id}>
        <Card
          title={project.title}
          person={project.person}
          time={project.time}
          progress={project.progress}
          status={project.status}
          labels={project.labels}
        />
      </div>
    ));
    return (
      <div key={month} className="App-month">
        <h2 className="App-monthTitle">{month}</h2>
        {cards}
      </div>
    );
  });

  const hasProjects = projects && projects.length;

  return (
    <div
      className="App"
      onDrop={e => handleDrop(onFileDrop, e)}
      onDragOver={handleDragOver}
    >
      <div className="App-sideMenu">
        <SideMenu onTabChange={onTabChange} selectedTab={selectedTab} />
      </div>
      <div className="App-header">
        <Header title={title} />
      </div>
      <div className="App-content">
        {errorMessage === undefined ? months : formatErrors(errorMessage)}
      </div>
      {hasProjects &&
        errorMessage === undefined &&
        <div className="App-footer">
          <Legend projects={projects} />
        </div>}
    </div>
  );
}

function splitIntoMonths(projects: Project[]) {
  const monthIndexes = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11];
  return monthIndexes
    .map(monthIndex =>
      projects.filter(project => moment(project.time).month() === monthIndex))
    .filter(month => month.length > 0)
    .map(month => ({
      month: moment(month[0].time).format("MMMM"),
      projects: month
    }));
}

function handleDragOver(e: React.DragEvent<HTMLDivElement>) {
  e.preventDefault();
}

function handleDrop(onFileDrop: (file: File) => void, e: React.DragEvent<HTMLDivElement>) {
  e.preventDefault();
  const dt = e.dataTransfer;
  if (dt.items) {
    const item = Array.prototype.find.call(
      dt.items,
      ({ kind }: DataTransferItem) => kind === "file"
    );
    if (item) onFileDrop(item.getAsFile());
  } else if (dt.files.length) {
    onFileDrop(dt.files[0]);
  }
}

function formatErrors(errorMessage: string | string[]) {
  if (typeof errorMessage === "string") return errorMessage;

  return (
    <ul>
      {errorMessage.map((message, i) => <li key={i}>{message}</li>)}
    </ul>
  );
}
