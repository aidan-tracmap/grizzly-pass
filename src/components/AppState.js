import React, { Component, PropTypes } from "react";
import { projects as demoProjects, labels as demoLabels } from "../demo-data";
import { generateLabelInfo } from "../label-utils";
import App from "./App";

export default class AppState extends Component {
  static propTypes = {
    db: PropTypes.func.isRequired,
    readFileAsText: PropTypes.func.isRequired
  };

  constructor(props) {
    super(props);
    this.handleFileDrop = this.handleFileDrop.bind(this);
    this.state = {
      title: "Demo Dashboard",
      projects: demoProjects,
      labels: demoLabels
    };
  }

  render() {
    const { title, projects, labels } = this.state;

    const labelInfo = generateLabelInfo(labels);
    const projectsWithLabels = projects.map(project => ({
      ...project,
      ...(!project.labels ? {} : { labels: getLabelInfo(labelInfo, project) })
    }));

    return (
      <App
        title={title}
        projects={projectsWithLabels}
        onFileDrop={this.handleFileDrop}
      />
    );
  }

  async handleFileDrop(file) {
    const asText = await this.props.readFileAsText(file);
    const { title, projects } = JSON.parse(asText);
    this.setState({
      projects,
      title
    });
  }
}

function getLabelInfo(labelInfo, project) {
  return project.labels.map(labelId => labelInfo[labelId]);
}
