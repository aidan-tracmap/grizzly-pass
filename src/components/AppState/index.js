// @flow

import type {
  Project,
  Label,
  LabelInfo,
  TabId,
  ProjectWithLabelInfo
} from "../../types";
import flow from "lodash/flow";
import React, { Component } from "react";
import {
  projects as demoProjects,
  labels as demoLabels
} from "../../demo-data";
import { generateLabelInfo } from "../../labels";
import { importFile } from "../../import";
import { ok, map, orElse } from "../../result";
import App from "../App";

type AppStateProps = {
  db: () => IDBDatabase,
  readFileAsText: (File) => Promise<string>
};

type AppStateState = {
  title: ?string,
  projects: Project[],
  labels: Label[],
  errorMessage: ?(string | string[]),
  selectedTab: ?TabId
};

export default class AppState
  extends Component<void, AppStateProps, AppStateState> {
  handleFileDrop: (File) => void;
  handleTabChange: (?TabId) => void;
  handleProjectsChange: (ProjectWithLabelInfo[]) => void;

  state = {
    title: "Demo Dashboard",
    projects: demoProjects,
    labels: demoLabels,
    errorMessage: null,
    selectedTab: null
  };

  constructor(props: AppStateProps) {
    super(props);
    this.handleFileDrop = this.handleFileDrop.bind(this);
    this.handleTabChange = this.handleTabChange.bind(this);
    this.handleProjectsChange = this.handleProjectsChange.bind(this);
  }

  render() {
    const { title, projects, labels, errorMessage, selectedTab } = this.state;

    const labelInfo = generateLabelInfo(labels);
    const projectsWithLabels = projects.map(project => ({
      id: project.id,
      title: project.title,
      person: project.person,
      time: project.time,
      progress: project.progress,
      status: project.status,
      labels: getLabelInfo(labelInfo, project)
    }));

    return (
      <App
        title={title}
        projects={projectsWithLabels}
        errorMessage={errorMessage}
        selectedTab={selectedTab}
        onFileDrop={this.handleFileDrop}
        onTabChange={this.handleTabChange}
        onProjectsChange={this.handleProjectsChange}
        editable={selectedTab === "edit"}
      />
    );
  }

  async handleFileDrop(file: File) {
    const asText = await this.props.readFileAsText(file);
    flow(
      importFile,
      map(fileData => ({ ...fileData, errorMessage: null })),
      orElse(errorMessage => ok({ errorMessage })),
      map(stateChange => this.setState(stateChange))
    )(asText);
  }

  handleTabChange(selectedTab: ?TabId) {
    this.setState({ selectedTab });
  }

  handleProjectsChange(projects: ProjectWithLabelInfo[]) {
    this.setState({
      projects: projects.map(project => ({
        id: project.id,
        title: project.title,
        person: project.person,
        time: project.time,
        progress: project.progress,
        status: project.status,
        labels: project.labels.map(({ id }) => id)
      }))
    });
  }
}

function getLabelInfo(
  labelInfo: { [string]: LabelInfo },
  project: Project
): LabelInfo[] {
  return project.labels.map(labelId => labelInfo[labelId]);
}
