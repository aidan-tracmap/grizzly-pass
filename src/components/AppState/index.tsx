import * as React from "react";
import {
  projects as demoProjects,
  labels as demoLabels
} from "../../demo-data";
import { generateLabelInfo } from "../../labels";
import { importFile } from "../../import";
import { Ok } from "../../result";
import { TabId, Status } from "../../types";
import App from "../App";
import { isArray, omit } from "lodash";

interface AppStateProps {
  db: () => IDBDatabase;
  readFileAsText: (file: File) => Promise<string>;
}

interface State {
  title?: string;
  projects: {
    id: string;
    title: string;
    person: string;
    time: string;
    progress: number;
    status: Status;
    labels?: string[]
  }[];
  labels: { title: string; id: string; }[];
  errorMessage?: string | string[];
  selectedTab?: TabId;
};

export default class AppState extends React.Component<AppStateProps, State> {
  constructor(props: AppStateProps) {
    super(props);
    this.handleFileDrop = this.handleFileDrop.bind(this);
    this.handleTabChange = this.handleTabChange.bind(this);
    this.state = {
      title: "Demo Dashboard",
      projects: demoProjects,
      labels: demoLabels,
    };
  }

  render() {
    const { title, projects, labels, errorMessage, selectedTab } = this.state;

    const labelInfo = generateLabelInfo(labels);
    const projectsWithLabels = projects.map(project => ({
      id: project.id,
      title: project.title,
      person: project.person,
      time: project.time,
      progress: project.progress,
      status: project.status,
      labels: getLabelInfo(labelInfo, project.labels)
    }));

    return (
      <App
        title={title}
        projects={projectsWithLabels}
        errorMessage={errorMessage}
        selectedTab={selectedTab}
        onFileDrop={this.handleFileDrop}
        onTabChange={this.handleTabChange}
      />
    );
  }

  async handleFileDrop(file: File) {
    const asText = await this.props.readFileAsText(file);
    importFile(asText)
      .map(fileData => ({ ...fileData, errorMessage: undefined }))
      .orElse(errorMessage => Ok({ errorMessage }))
      .map(stateChange => (this.setState as any)(stateChange));
  }

  handleTabChange(selectedTab: TabId | null) {
    this.setState({ selectedTab: selectedTab === null ? undefined : selectedTab });
  }
}

function getLabelInfo(
  labelInfo: Record<string, { id: string; initial: string; colour: string; title: string; }>,
  labels?: string[]
): { id: string; initial: string; colour: string; title: string; }[] | undefined {
  if (isArray(labels))
    return labels.map(labelId => labelInfo[labelId]);
  else
    return undefined;
}
