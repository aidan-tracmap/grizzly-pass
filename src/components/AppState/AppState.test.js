import React from "react";
import { shallow } from "enzyme";
import omit from "lodash/omit";
import AppState from "../AppState";
import {
  projects as demoProjects,
  labels as demoLabels
} from "../../demo-data";
import { tabIds } from "../../types";
import App from "../App";

const jsonFile = [
  {
    id: "1",
    title: "Coffee Swirl",
    person: "Joe Lemon",
    date: "2017-03-15",
    progress: 13,
    health: "ontrack"
  },
  {
    id: "2",
    title: "Coffee Swirl",
    person: "Joe Lemon",
    date: "2017-03-15",
    progress: 13,
    health: "ontrack",
    tags: ["Apple"]
  }
];

function renderAppState(file = jsonFile) {
  const readFileAsText = jest.fn(() => JSON.stringify(jsonFile));
  return shallow(<AppState db={() => {}} readFileAsText={readFileAsText} />);
}

const db = jest.fn();
const readFileAsText = jest.fn(() => JSON.stringify(jsonFile));
const appState = shallow(<AppState db={db} readFileAsText={readFileAsText} />);

it("uses the demo data as it's default projects state", () => {
  expect(appState.state("title")).toBe("Demo Dashboard");
  expect(appState.state("projects")).toMatchSnapshot();
  expect(appState.state("labels")).toMatchSnapshot();
});

it("has no error message by default", () => {
  expect(appState.state("errorMessage")).toBeFalsy();
});

it("passes the projects (with embedded label info) to it's child <App />", () => {
  appState.setState({
    projects: demoProjects,
    labels: demoLabels
  });
  expect(appState.find(App).prop("projects")).toEqual([
    {
      id: "1",
      title: "Coffee Swirl",
      person: "Joe Lemon",
      time: "2017-03-15T10:54:04.445Z",
      progress: 13,
      status: "ontrack",
      labels: [
        {
          id: "3",
          initial: "A",
          colour: "#EF5350",
          title: "Apple"
        }
      ]
    },
    {
      id: "2",
      title: "Rake Twister",
      person: "Alex Apple",
      time: "2017-04-12T10:54:04.445Z",
      progress: 50,
      status: "onhold",
      labels: [
        {
          id: "3",
          initial: "A",
          colour: "#EF5350",
          title: "Apple"
        },
        {
          id: "4",
          initial: "O",
          colour: "#7E57C2",
          title: "Orange"
        }
      ]
    }
  ]);
});

it("passes no label info with projects (if there's no labels) to it's child <App />", () => {
  appState.setState({
    projects: demoProjects.map(project => omit(project, "labels"))
  });
  expect(appState.find(App).prop("projects")).toEqual([
    {
      id: "1",
      title: "Coffee Swirl",
      person: "Joe Lemon",
      time: "2017-03-15T10:54:04.445Z",
      progress: 13,
      status: "ontrack"
    },
    {
      id: "2",
      title: "Rake Twister",
      person: "Alex Apple",
      time: "2017-04-12T10:54:04.445Z",
      progress: 50,
      status: "onhold"
    }
  ]);
});

it("passes the title state to it's child <App />", () => {
  const title = appState.state("title");
  expect(appState.find(App).prop("title")).toBe(title);
});

it("passess the errorMessage state to it's child <App />", () => {
  appState.setState({ errorMessage: "Error message" });
  expect(appState.find(App).prop("errorMessage")).toBe("Error message");
});

it("defaults to no tabs selected", () => {
  const appState = renderAppState();
  expect(appState.find(App).prop("selectedTab")).toBe(null);
});

it("passes the selectedTab state to it's child <App />", () => {
  const appState = renderAppState();
  appState.setState({ selectedTab: tabIds[0] });
  expect(appState.find(App).prop("selectedTab")).toBe(tabIds[0]);
});

it("updates the selectedTab state", () => {
  const appState = renderAppState();
  expect(appState.find(App).prop("selectedTab")).not.toBe(tabIds[0]);
  appState.instance().handleTabChange(tabIds[0]);
  expect(appState.find(App).prop("selectedTab")).toBe(tabIds[0]);
});

it("updates the projects state if the dropped file is json", async function() {
  const mockFile = { MOCK: "FILE" };
  await appState.instance().handleFileDrop(mockFile);
  expect(readFileAsText).toHaveBeenCalledWith(mockFile);
  expect(appState.state("errorMessage")).toBeFalsy();
  expect(appState.state("title")).toBe(null);
  expect(appState.state("projects").length).toEqual(2);
  expect(appState.state("labels").length).toBe(1);
});

it("sets the errorMessage state if the dropped file is invalid", async function() {
  const mockFile = { MOCK: "FILE" };
  readFileAsText.mockImplementationOnce(() =>
    JSON.stringify("this is not a projects file"));
  await appState.instance().handleFileDrop(mockFile);
  expect(readFileAsText).toHaveBeenCalledWith(mockFile);
  expect(appState.state("errorMessage")).toMatchSnapshot();
});
