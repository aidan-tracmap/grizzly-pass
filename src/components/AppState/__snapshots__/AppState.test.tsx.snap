exports[`test sets the errorMessage state if the dropped file is invalid 1`] = `"File isn\'t a JSON array"`;

exports[`test uses the demo data as it's default projects state 1`] = `
Array [
  Object {
    "id": "1",
    "labels": Array [
      "3",
    ],
    "person": "Joe Lemon",
    "progress": 13,
    "status": "ontrack",
    "time": "2017-03-15T10:54:04.445Z",
    "title": "Coffee Swirl",
  },
  Object {
    "id": "2",
    "labels": Array [
      "3",
      "4",
    ],
    "person": "Alex Apple",
    "progress": 50,
    "status": "onhold",
    "time": "2017-04-12T10:54:04.445Z",
    "title": "Rake Twister",
  },
]
`;

exports[`test uses the demo data as it's default projects state 2`] = `
Array [
  Object {
    "id": "3",
    "title": "Apple",
  },
  Object {
    "id": "4",
    "title": "Orange",
  },
]
`;
